staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness_test.go

```diff
@@ -0,0 +1,261 @@
+	switch t.mockDecision {
+	case decisionNoQueuingExecute:
+		execFn()
+	case decisionQueuingExecute:
+		queueNoteFn(true)
+		t.postEnqueue()
+		queueNoteFn(false)
+		t.postDequeue()
+		execFn()
+	case decisionCancelWait:
+		queueNoteFn(true)
+		t.postEnqueue()
+		queueNoteFn(false)
+		t.postDequeue()
+	case decisionReject:
+		return
+	}
+}
+
+func (t fakeApfFilter) Run(stopCh <-chan struct{}) error {
+	return nil
+}
+
+func (t fakeApfFilter) Install(c *mux.PathRecorderMux) {
+}
+
+func newApfServer(decision int, t *testing.T) *httptest.Server {
+	requestInfoFactory := &apirequest.RequestInfoFactory{APIPrefixes: sets.NewString("apis", "api"), GrouplessAPIPrefixes: sets.NewString("api")}
+	longRunningRequestCheck := BasicLongRunningRequestCheck(sets.NewString("watch"), sets.NewString("proxy"))
+
+	apfHandler := WithPriorityAndFairness(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if decision == decisionCancelWait {
+			t.Errorf("execute should not be invoked")
+		}
+		if decision != decisionSkipFilter && atomicReadOnlyExecuting != 1 {
+			t.Errorf("Wanted %d requests executing, got %d", 1, atomicReadOnlyExecuting)
+		}
+	}), longRunningRequestCheck, fakeApfFilter{
+		mockDecision: decision,
+		postEnqueue: func() {
+			if atomicReadOnlyWaiting != 1 {
+				t.Errorf("Wanted %d requests in queue, got %d", 1, atomicReadOnlyWaiting)
+			}
+		},
+		postDequeue: func() {
+			if atomicReadOnlyWaiting != 0 {
+				t.Errorf("Wanted %d requests in queue, got %d", 0, atomicReadOnlyWaiting)
+			}
+		},
+	})
+
+	handler := apifilters.WithRequestInfo(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		r = r.WithContext(apirequest.WithUser(r.Context(), &user.DefaultInfo{
+			Groups: []string{user.AllUnauthenticated},
+		}))
+		apfHandler.ServeHTTP(w, r)
+		if atomicReadOnlyExecuting != 0 {
+			t.Errorf("Wanted %d requests executing, got %d", 0, atomicReadOnlyExecuting)
+		}
+	}), requestInfoFactory)
+
+	apfServer := httptest.NewServer(handler)
+	return apfServer
+}
+
+func TestApfSkipLongRunningRequest(t *testing.T) {
+	epmetrics.Register()
+
+	server := newApfServer(decisionSkipFilter, t)
+	defer server.Close()
+
+	if err := expectHTTPGet(fmt.Sprintf("%s/api/v1/namespaces?watch=true", server.URL), http.StatusOK); err != nil {
+		// request should not be rejected
+		t.Error(err)
+	}
+}
+
+func TestApfRejectRequest(t *testing.T) {
+	epmetrics.Register()
+
+	server := newApfServer(decisionReject, t)
+	defer server.Close()
+
+	if err := expectHTTPGet(fmt.Sprintf("%s/api/v1/namespaces/default", server.URL), http.StatusTooManyRequests); err != nil {
+		t.Error(err)
+	}
+
+	checkForExpectedMetricsWithRetry(t, []string{
+		"apiserver_request_terminations_total",
+		"apiserver_dropped_requests_total",
+	})
+}
+
+func TestApfExemptRequest(t *testing.T) {
+	epmetrics.Register()
+	fcmetrics.Register()
+
+	// wait the first sampleAndWaterMark metrics to be collected
```

Comment:
This comment is misleading, as @ahg-g 's remark below indicates.
Perhaps something along the following lines would be better.
> Wait for at least one sampling window to pass since creation of `k8s.io/apiserver/pkg/util/flowcontrol/metrics.ReadWriteConcurrencyObserverPairGenerator`,
> so that an observation will cause some data to go into the Prometheus metrics.

Reply:
Is there a deterministic way to check that a sampling window has passed? is there an API we can call to tell us that?

Reply:
We can deterministically know that at least one sampling window has passed by waiting for at least the duration of a sampling window.  The existing test func does this.

Reply:
Updated comment. Much appreciated your suggestions. @MikeSpreitzer